File: dynamic-programming/longest_increasing_subsequence.py
# ---
# title: Longest Increasing Subsequence (LIS)
# tags: [CodingChallenges, Python, DynamicProgramming]
# created: '2022-11-07'
# ---

# ## Problem Description

# The longest increasing subsequence problem is to find a subsequence of a given sequence in which the subsequence's elements are in sorted order, lowest to highest, and in which the subsequence is as long as possible. 
# This subsequence is not necessarily contiguous or unique.

# ## Solution

# There are two popular ways to solve this problem:

# 1. Using dynamic programming with time complexity O(N^2)
# 2. Using segment tree with time complexity O(NlogN) (we will cover this after learning segment trees)

# ### Dynamic Programming Approach

# In this approach, we create an array D where each index represents the length of the longest increasing subsequence ending at that index. 
# For each element at index i, we check all previous elements and update D[i] accordingly.


# ```python
N = int(input())
A = list(map(int,input().split()))
D = [0] * N

for i in range(N):
  D[i] = 1
  for j in range(i):
    if A[i] > A[j]:
      D[i] = max(D[i],D[j]+1)
    
print(max(D))
# ```


# ### Segment Tree Approach (To be continued...)

# ```python
# from itertools import chain
# n = int(input())
# a = list(map(int,input().split()))
# dp = [[0,0] for _ in range(n)]

# for i in range(n):
#   dp[i][0] = 1
#   dp[i][1] = 1
#   for j in range(i):
#     if a[i] > a[j]:
#       dp[i][0] = max(dp[i][0],dp[j][0]+1)
#     if a[j] > a[i]:
#       dp[i][1] = max(dp[i][1],dp[j][0]+1,dp[j][1]+1)

# print(max(list(chain(*dp)))) 
# ```

# In the segment tree approach, we will sort the array and apply dynamic programming on it using segment trees which will reduce our time complexity to O(NlogN). 
# However, understanding of segment trees is required before implementing this solution.

# ### Applying LIS on pairs

# This approach is an extension of LIS where instead of working on single numbers we work on pairs.
# Here first we sort our pairs based on first element then apply LIS on second element.

# ```python
# n = int(input())
# a = [list(map(int,input().split())) for _ in range(n)]
# a.sort()
# dp = [0] * n

# for i in range(n):
#   dp[i] = 1
#   for j in range(i):
#     if a[i][1] > a[j][1]:
#       dp[i] = max(dp[i],dp[j]+1)

# print(n-max(dp))
# ```

File: dynamic-programming/longestIncreasingSubsequence.js
// ---
// title: Longest Increasing Subsequence (LIS)
// tags: [CodingChallenges, JavaScript, DynamicProgramming, LIS]
// # created: '2022-11-07'
// ---

// ## Problem Description

// Given an array of integers, find the length of the longest increasing subsequence (LIS). An increasing subsequence is a subsequence where all the elements are in sorted order, lowest to highest.

// ## Solution

// The solution uses dynamic programming to keep track of the longest increasing subsequences ending at each index. For each element in the array, we iterate through all previous elements and if current element is greater than any previous one and maximum length at that previous index is more than current maximum length then update current maximum length.

// Pseudocode:
// 1. Start from first index.
// 2. If current element is larger than its preceding elements, add 1.
// 3. When comparing with preceding elements and if current element is larger, add the largest LIS value among those preceding elements.

// Here's how this looks in JavaScript:

// ```javascript
const LIS = function (arr) {
  const n = arr.length;
  const lis = Array(n).fill(1);

  for (let i = 1; i < n; i++) {
    let max = 0;
    for (let j = 0; j < i; j++) {
      if (arr[j] < arr[i] && max < lis[j]) {
        max = lis[j];
      }
    }
    lis[i] += max;
  }

  return Math.max(...lis);
};

let output = LIS([3, 2]);
console.log(output); // --> 1

output = LIS([3, 10, 2, 1, 20]);
console.log(output); // --> 3

output = LIS([10, 20, '40', 25, '30', '50', 32, '70', 85]);
console.log(output); // -->7

output = LIS(['7', '2', '6', '4', '5', '1', '3']);
console.log(output); // -->3

output = LIS(['7', '2', '5', '3', '4', '9', '1', '10', '2']);
console.log(output); //-->5

output = LIS(['2', '5', '3', '7', '11', '8', '10', '13', '6']);
console.log(output); //-->6
